multidialer.go

```diff
@@ -41,49 +43,48 @@ func (d *multiDialer) DialContext(ctx context.Context, network, addr string) (ne
 		return nil, fmt.Errorf("undialable IP: %s", ip)
 	}
 
-	router, err := netroute.New()
-	if err != nil {
-		return nil, err
-	}
-
-	_, _, preferredSrcIP, err := router.Route(ip)
-	if err != nil {
-		return nil, err
-	}
-
-	loopbackCandidates := make([]*net.TCPAddr, 0)
-	unspecifiedCandidates := make([]*net.TCPAddr, 0)
-	existingPort := 0
-	for _, optAddr := range d.listeningAddresses {
-		if optAddr.IP.Equal(preferredSrcIP) {
-			return reuseDial(ctx, optAddr, network, addr)
-		}
-		if optAddr.IP.IsLoopback() {
-			loopbackCandidates = append(loopbackCandidates, optAddr)
-		} else if optAddr.IP.IsGlobalUnicast() && existingPort == 0 {
-			existingPort = optAddr.Port
-		} else if optAddr.IP.IsUnspecified() {
-			unspecifiedCandidates = append(unspecifiedCandidates, optAddr)
+	if router, err := netroute.New(); err == nil {
+		if _, _, preferredSrc, err := router.Route(ip); err == nil {
+			for _, optAddr := range d.listeningAddresses {
+				if optAddr.IP.Equal(preferredSrc) {
+					return reuseDial(ctx, optAddr, network, addr)
+				}
+			}
 		}
 	}
-	if ip.IsLoopback() && len(loopbackCandidates) > 0 {
-		return reuseDial(ctx, randAddr(loopbackCandidates), network, addr)
+
+	if ip.IsLoopback() && len(d.loopback) > 0 {
+		return reuseDial(ctx, randAddr(d.loopback), network, addr)
 	}
```

Comment:
nit: I think we can move this up and skip netroute entirely for localhost.

Reply:
if there are multiple loopback entries, netroute can do better than random.